internal/component/pyroscope/relabel/relabel.go

```diff
@@ -0,0 +1,337 @@
+type Exports struct {
+	Receiver pyroscope.Appendable `alloy:"receiver,attr"`
+	Rules    alloy_relabel.Rules  `alloy:"rules,attr"`
+}
+
+// Component implements the pyroscope.relabel component.
+type Component struct {
+	opts         component.Options
+	metrics      *metrics
+	mut          sync.RWMutex
+	rcs          []*relabel.Config
+	fanout       *pyroscope.Fanout
+	cache        *lru.Cache
+	maxCacheSize int
+	exited       atomic.Bool
+}
+
+var (
+	_ component.Component = (*Component)(nil)
+)
+
+// New creates a new pyroscope.relabel component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	cache, err := lru.New(args.MaxCacheSize)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Component{
+		opts:         o,
+		metrics:      newMetrics(o.Registerer),
+		cache:        cache,
+		maxCacheSize: args.MaxCacheSize,
+	}
+
+	c.fanout = pyroscope.NewFanout(args.ForwardTo, o.ID, o.Registerer)
+
+	o.OnStateChange(Exports{
+		Receiver: c,
+		Rules:    args.RelabelConfigs,
+	})
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+func (c *Component) Run(ctx context.Context) error {
+	defer c.exited.Store(true)
+	<-ctx.Done()
+	return nil
+}
+
+func (c *Component) Update(args component.Arguments) error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	newArgs := args.(Arguments)
+	newRCS := alloy_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelConfigs)
+
+	// If relabeling rules changed, purge the cache
+	if relabelingChanged(c.rcs, newRCS) {
+		level.Debug(c.opts.Logger).Log("msg", "received new relabel configs, purging cache")
+		c.cache.Purge()
+		c.metrics.cacheSize.Set(0)
+	}
+
+	if newArgs.MaxCacheSize != c.maxCacheSize {
+		evicted := c.cache.Resize(newArgs.MaxCacheSize)
+		if evicted > 0 {
+			level.Debug(c.opts.Logger).Log("msg", "resizing cache led to evicting items", "evicted_count", evicted)
+		}
+		c.maxCacheSize = newArgs.MaxCacheSize
+	}
+
+	c.rcs = newRCS
+	c.fanout.UpdateChildren(newArgs.ForwardTo)
+
+	c.opts.OnStateChange(Exports{
+		Receiver: c,
+		Rules:    newArgs.RelabelConfigs,
+	})
+
+	return nil
+}
+
+type ProfileDroppedError struct{}
+
+func (e *ProfileDroppedError) Error() string {
+	return "profile dropped by relabel rules"
+}
+
+func (c *Component) Append(ctx context.Context, lbls labels.Labels, samples []*pyroscope.RawSample) error {
+	if c.exited.Load() {
+		return fmt.Errorf("%s has exited", c.opts.ID)
+	}
+
+	c.mut.RLock()
+	defer c.mut.RUnlock()
+
+	newLabels, keep, err := c.relabel(lbls)
+	if err != nil {
+		return err
+	}
+	if !keep {
+		return &ProfileDroppedError{}
+	}
+
+	return c.fanout.Appender().Append(ctx, newLabels, samples)
+}
+
+func (c *Component) AppendIngest(ctx context.Context, profile *pyroscope.IncomingProfile) error {
+	if c.exited.Load() {
+		return fmt.Errorf("%s has exited", c.opts.ID)
+	}
+
+	c.mut.RLock()
+	defer c.mut.RUnlock()
+
+	lbls, err := extractLabelsFromIncomingProfile(profile)
```

Comment:
I think maybe it is better to modify `*pyroscope.IncomingProfile` to have `labels.Label` in the extracted form and then at very last minute (e.g. in `pyroscope.write`, convert it back to the url query parameter.)

Reply:
Done in 8241435c2a2574235b672aae2568f87eac1fab84